File: Exercise_2.cpp
// Time Complexity                                  : PUSH    - O(1)
//                                                    POP     - O(1)
//                                                    TOP     - O(1)
//                                                    GETMIN  - O(1)
// 
// Space Complexity                                 : O(N) = O(N) - storing key, value and next for each value.
// Did this code successfully run on Leetcode       : Yes.
// Any problem you faced while coding this          : No.

#include<bits/stdc++.h>
using namespace std;

class MinStack {
public:
    stack<int> st;
    int mn;
    MinStack() {
        
    }
    
    void push(int val) {
        
        if(st.empty()) {
            st.push(INT_MAX);
            mn = val;
            st.push(val);
        } else {
            if(val <= mn) {
                st.push(mn);
                mn = val;
            }
            st.push(val);
        }
    }
    
    void pop() {
        int m = st.top();
        if(m == mn) {
            st.pop();
            mn = st.top();
        }
        st.pop();  
    }
    
    int top() {
        return st.top();
    }
    
    int getMin() {
        return mn;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: Exercise_2.js
class MinStack {
    constructor() {
        this.st = new Array()
        this.minSt = new Array()
        this.counter = -1
        this.minCounter = 0
        this.min = Number.MAX_SAFE_INTEGER;
        this.minSt.push(this.min)
    }

    push(val) {
        const calMin = Math.min(this.min, val)
        this.st.push(val)
        this.minSt.push(calMin)
        this.counter = this.counter + 1
        this.minCounter = this.minCounter + 1
        this.min = calMin
    }

    pop() {
        this.st.pop()
        this.minSt.pop()
        this.counter = this.counter - 1
        this.minCounter = this.minCounter - 1
        this.min = this.minSt[this.minCounter]
    }

    top() {
        return this.st[this.counter]
    }

    getMin() {
        return this.min
    }
}